partiql-lang/src/main/kotlin/org/partiql/lang/eval/EvaluatingCompiler.kt

```diff
@@ -2159,6 +2206,138 @@ internal open class EvaluatingCompiler(
             CompiledGroupByItem(alias.text.exprValue(), uniqueName, compileAstExpr(it.expr))
         }
 
+    /**
+     * Represents an instance of a compiled `EXCLUDE` expression. Notably, this expr will have redundant steps removed.
+     */
+    internal data class CompiledExcludeExpr(val root: PartiqlAst.Identifier, val exclusions: RemoveAndOtherSteps)
+    /**
+     * Represents all the exclusions at the current level and other nested levels.
+     */
+    internal data class RemoveAndOtherSteps(val remove: Set<PartiqlAst.ExcludeStep>, val steps: Map<PartiqlAst.ExcludeStep, RemoveAndOtherSteps>) {
```

Comment:
Could you please help me understand how the steps are modeled? Why is there a set of steps and then a map from steps to more? This seems like an odd way to model what looks to be a cluster of trees

Reply:
Understanding this will help be better see what's going on in compileExcludeClause

Reply:
For example, let's say we have exclude paths (ignoring the exclude path root) of
```
      a.b,     -- exclude path 1
      x.y.z1,  -- exclude path 2
      x.y.z2   -- exclude path 3
      ^ ^ ^
Level 1 2 3
```

These exclude paths would be converted to the following in `RemoveAndOtherSteps`.
```Kotlin
// For demonstration purposes, the syntax '<string>' corresponds to the exclude tuple attribute step of <string>
RemoveAndOtherSteps(                   // Level 1 (no exclusions at level 1)
    remove = emptySet(),
    steps = mapOf(
        'a' to RemoveAndOtherSteps(    // Level 2 for paths that have `'a'` in Level 1
            remove = setOf('b'),       // path `a.b` has final step at level 2
            steps = emptyMap()
        ),
        'x' to RemoveAndOtherSteps(    // Level 2 for paths that have `'x'` in Level 1
            remove = emptySet(),
            steps = mapOf(
                'y' to RemoveAndOtherSteps(     // Level 3 for paths that have `'y'` in Level 2 and `'x'` in Level 1
                    remove = setOf('z1', 'z2'), // paths `x.y.z1` and `x.y.z2` have final step at level 3
                    steps = emptyMap()
                )
            )
        ),
    )
)
```

Some of the other `EXCLUDE` path merging/subsumption tests shows the relationship between `EXCLUDE` paths and this data structure: https://github.com/partiql/partiql-lang-kotlin/blob/2d05d8a13237ed0a0ed1e1d5cec53171ca8d8b28/partiql-lang/src/test/kotlin/org/partiql/lang/eval/EvaluatingCompilerExcludeTests.kt#L841-L1074C14.

Reply:
I agree that the naming/factoring of `RemoveAndOtherSteps` could be initially confusing. Since this is currently internal, I'll leave it as it is. I plan to take another look at refactoring this as part of porting `EXCLUDE` to the plan compiler in #1280.